arch/sparc/core/trap_table_svt.S

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2023 Frontgrade Gaisler AB
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This file contains the trap entry for SPARC operating with
+ * single-vector trap model, defined in SPARC V8E. The processor
+ * redirects execution to a single entry on any trap event. From
+ * there, two levels of look-up tables are used to find the trap
+ * handler.
+ *
+ * - Execution time is constant.
+ * - Condition flags are not modified.
+ * - Provides handler with PSR in l0, TBR in l6
+ * - This SVT implementation is less than 400 bytes long. (An MVT
+ *   table is always 4096 bytes long.)
+ *
+ * See trap_table_mvt.S for information about SPARC trap types.
+ */
+
+#include <zephyr/toolchain.h>
+#include <zephyr/linker/sections.h>
+#include <zephyr/arch/sparc/sparc.h>
+
+#ifdef CONFIG_IRQ_OFFLOAD
+ #define IRQ_OFFLOAD_HANDLER    __sparc_trap_irq_offload
+#else
+ #define IRQ_OFFLOAD_HANDLER    __sparc_trap_fault
+#endif
+
+GTEXT(__sparc_trap_table)
+GTEXT(__start)
+
+SECTION_SUBSEC_FUNC(TEXT, traptable, __sparc_trap_table)
+__start:
+	rd	%psr, %l0
+	mov	%tbr, %l6
+
+	and	%l6, 0xf00, %l7
+	srl	%l7, 6, %l7
+	set	__sparc_trap_table_svt_level0, %l4
+	ld	[%l4 + %l7], %l4
+
+	and	%l6, 0x0f0, %l7
+	srl	%l7, 2, %l7
+	ld	[%l4 + %l7], %l4
+
+	srl	%l6, 4, %l3
+	jmp	%l4
+	 and	%l3, 0xf, %l3 /* Interrupt level */
+
+__sparc_trap_svt_in_trap:
+	ta	0x00
+	nop
+
+SECTION_VAR(RODATA, __sparc_trap_table_svt_tables)
+	.align	4
+__sparc_trap_table_svt_level0:
+	.word	__sparc_trap_table_svt_00
+	.word	__sparc_trap_table_svt_10
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_80
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
+	.word	__sparc_trap_table_svt_allbad
```

Comment:
Nitpick as this is all down in the arch code, but building giant tables in assembly seems maybe a little questionable?  Are you sure this wouldn't be much clearer (and probably less error prone) as a table of pointers in C code instead?

Or even automatically generated, since it looks like it has complicated structure (tables are sequential, header contains pointers into the other tables, etc... again I don't know the rules because it's not my arch).

Reply:
The tables were put in the same file as the dispatch code to have the implementation self-contained and to keep complexity down.

Also, with a self-contained `trap_table_svt.S`, the relation to `trap_table_mvt.S` is emphasized as only one or the other is selected at build time. Even though these two `.S` files look different, they achieve the same thing: redirect execution to the target low-level assembly handler at trap time.

There is no support for dynamically updating the trap tables or similar in this PR. That could be an interesting extension (like dynamically change bus error handler or temporary custom instruction emulation) if use cases arise. In that case, I think it could be helpful to also turn the tables into a C code description.